pandas/core/indexing.py

```diff
@@ -1690,18 +1690,42 @@ def _setitem_with_indexer(self, indexer, value):
                     sub_indexer = list(indexer)
                     multiindex_indexer = isinstance(labels, ABCMultiIndex)
                     # TODO: we are implicitly assuming value.columns is unique
+                    unique_cols = value.columns.is_unique
+
+                    if not unique_cols and value.columns.equals(self.obj.columns):
+                        # We assume we are already aligned, see
+                        # test_iloc_setitem_frame_duplicate_columns_multiple_blocks
+                        for loc in ilocs:
+                            item = item_labels[loc]
+                            if item in value:
+                                sub_indexer[info_axis] = item
+                                v = self._align_series(
+                                    tuple(sub_indexer),
+                                    value.iloc[:, loc],
+                                    multiindex_indexer,
+                                )
+                            else:
+                                v = np.nan
 
-                    for loc in ilocs:
-                        item = item_labels[loc]
-                        if item in value:
-                            sub_indexer[info_axis] = item
-                            v = self._align_series(
-                                tuple(sub_indexer), value[item], multiindex_indexer
-                            )
-                        else:
-                            v = np.nan
+                            self._setitem_single_column(loc, v, pi)
 
-                        self._setitem_single_column(loc, v, pi)
+                    elif not unique_cols:
+                        raise ValueError(
+                            "Setting with non-unique columns is not allowed."
+                        )
+
+                    else:
+                        for loc in ilocs:
+                            item = item_labels[loc]
+                            if item in value:
+                                sub_indexer[info_axis] = item
+                                v = self._align_series(
+                                    tuple(sub_indexer), value[item], multiindex_indexer
+                                )
+                            else:
```

Comment:
what is the end goal here?

Reply:
i dont understand the question.  1718-1728 is just an indented version of what we have in the status quo

Reply:
it looks like you separated out the cases and duplicated the code. So is this making it easier to understand, later refactoring, ?

Reply:
its similar but not identical.  this chunk is aligning by label while the chunk above is aligning positionally.

yes, i do hope to refactor this method before too long

Reply:
ok cool

> yes, i do hope to refactor this method before too long

yeah these indexing routines defintely need some TLC